test/keyboardNavigation.test.tsx

```diff
@@ -91,38 +78,145 @@ test('at-bounds basic keyboard navigation', () => {
   validateCellPosition(6, 99);
   userEvent.keyboard('{ctrl}{home}');
   validateCellPosition(0, 0);
-  validateCheckboxHasFocus();
   userEvent.keyboard('{home}');
-  validateCheckboxHasFocus();
   validateCellPosition(0, 0);
   userEvent.keyboard('{ctrl}{home}');
-  validateCheckboxHasFocus();
   validateCellPosition(0, 0);
   fireEvent.keyDown(document.activeElement!, { key: 'PageUp' });
   validateCellPosition(0, 0);
 
-  // shift+tab tabs out of the grid
+  // tab at the end of a row selects the first cell on the next row
+  userEvent.keyboard('{end}');
+  userEvent.tab();
+  validateCellPosition(0, 1);
+
+  // shift tab should select the last cell of the previous row
+  userEvent.tab({ shift: true });
+  validateCellPosition(6, 0);
+});
+
+test('grid enter/exit', () => {
+  setup({ columns, rows: Array(5), summaryRows });
+
+  // no initial selection
+  expect(getSelectedCell()).not.toBeInTheDocument();
+
+  // tab into the grid
+  userEvent.tab();
+  validateCellPosition(0, 0);
+
+  // shift+tab tabs out of the grid if we are at the first cell
   userEvent.tab({ shift: true });
   expect(document.body).toHaveFocus();
 
-  // tab at the end of a row selects the first cell on the next row
   userEvent.tab();
-  userEvent.keyboard('{end}');
+  validateCellPosition(0, 0);
+
+  userEvent.keyboard('{arrowdown}');
+  userEvent.keyboard('{arrowdown}');
+  validateCellPosition(0, 2);
+
+  // tab should select the last selected cell
+  // click outside the grid
+  userEvent.click(document.body);
   userEvent.tab();
-  validateCellPosition(0, 1);
+  userEvent.keyboard('{arrowdown}');
+  validateCellPosition(0, 3);
+
+  // shift+tab should select the last selected cell
+  userEvent.click(document.body);
+  userEvent.tab({ shift: true });
+  userEvent.keyboard('{arrowup}');
+  validateCellPosition(0, 2);
 
-  // tab at the end of the grid tabs out of the grid
+  // tab tabs out of the grid if we are at the last cell
   userEvent.keyboard('{ctrl}{end}');
   userEvent.tab();
   expect(document.body).toHaveFocus();
 });
 
+test('navigation with focusable formatter', () => {
+  setup({ columns, rows: Array(1), summaryRows });
+  userEvent.tab();
+  validateCellPosition(0, 0);
+
+  // cell should not set tabIndex to 0 if it contains a focusable formatter
+  expect(getSelectedCell()).toHaveAttribute('tabIndex', '-1');
+  const checkbox = getSelectedCell()!.querySelector('input');
+  expect(checkbox).toHaveFocus();
+  expect(checkbox).toHaveAttribute('tabIndex', '0');
+
+  userEvent.tab();
+  validateCellPosition(1, 0);
+  // cell should set tabIndex to 0 if it does not have focusable formatter
+  expect(getSelectedCell()).toHaveAttribute('tabIndex', '0');
+});
+
+test('navigation when header and summary rows have focusable elements', () => {
```

Comment:
This would have failed with this [fix](https://github.com/adazzle/react-data-grid/pull/2588)